hibernate-core/src/main/java/org/hibernate/internal/ScrollableResultsStreamSupport.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.internal;
+
+import org.hibernate.ScrollableResults;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+import java.util.stream.Stream;
+
+import static java.util.Spliterators.spliteratorUnknownSize;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * Utilities for obtaining {@link Stream} with an underlying {@link ScrollableResults}
+ *
+ * @author Christophe Taret
+ */
+public class ScrollableResultsStreamSupport {
+
+	/**
+	 * @see Spliterator for description
+	 */
+	private static final int DEFAULT_CHARACTERISTICS =
+			Spliterator.NONNULL |
+					Spliterator.IMMUTABLE;
+
+	private ScrollableResultsStreamSupport() {
+	}
+
+	/**
+	 * @param scrollableResults the {@link ScrollableResults}
+	 * @param type              the class of the {@link Stream}
+	 * @param <T>               the type of the {@link Stream}
+	 * @return a {@link Stream} with values from the {@link ScrollableResults}
+	 */
+	public static <T> Stream<T> asStream(ScrollableResults scrollableResults, Class<T> type) {
+		return asStream( scrollableResults, type, DEFAULT_CHARACTERISTICS, false );
+	}
+
+	/**
+	 * @param scrollableResults the {@link ScrollableResults}
+	 * @param type              the class of the {@link Stream}
+	 * @param <T>               the type of the {@link Stream}
+	 * @return a parallel {@link Stream} with values from the {@link ScrollableResults}
+	 */
+	public static <T> Stream<T> asParallelStream(ScrollableResults scrollableResults, Class<T> type) {
+		return asStream( scrollableResults, type, DEFAULT_CHARACTERISTICS, true );
+	}
+
+	private static <T> Stream<T> asStream(
+			ScrollableResults scrollableResults,
+			Class<T> type,
+			int characteristics,
+			boolean parallel) {
+		// build the stream whether parallel or not, with all required information
+		return stream(
+				spliteratorUnknownSize(
+						new ScrollableResultIterator<>(scrollableResults, type),
+						characteristics),
+				parallel);
```

Comment:
You should add an `.onClose(scrollableResults::close)` when constructing the stream, to make sure the `ScrollableResults` will be closed when the stream is closed:
```
return stream(
                spliteratorUnknownSize(
                        new ScrollableResultIterator<>(scrollableResults, type),
                        characteristics),
                parallel)
                .onClose(scrollableResults::close);
```